[1단계 - 자동차 경주 구현] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request:
안녕하세요. 
페어프로그래밍을 진행하며 몇 가지 궁금한 점이 있어 정리해 보았습니다.

1. 생성자에서 값을 초기화 할때도 상수를 사용하는 것이 좋나요?
  Car객체에 `private static final int INITIAL_POSITION = 0;`와 같이 자동차 위치를 0이라는 수로 초기화를 할 때 상수로 초기화하지 않아도 코드를 보면 해당 코드가 자동차의 초기 위치를 설정하는 것을 알 수 있는데 이러한 값들도 상수를 지정해주는게 좋은지 알고 싶습니다. (상수 정의의 기준을 알고 싶습니다.)

2. getter를 반드시 없애야 하나요?
  https://techcourse.woowahan.com/s/NASxoOCp/ls/RCk0xCYe를 보면 "의식적인 연습 2 - 객체지향 생활 체조 원칙"에서 "게터를 쓰지 않는다"라는 규칙이 있습니다. 그런데 프리코드에서 받은 피드백 중 "뷰와 컨트롤러를 하나의 클래스에 넣지 않는다, 클래스는 하나의 일만 해야 한다 "는 원칙도 있습니다. 현재 코드에서 Output 클래스에서 결과를 출력하기 위해 Car에서 getName()과 getPosition()을 사용하고 있습니다. 이런 코드를 작성한 이윤는 프리코스 피트백을 만족하기 위해서입니다. 만약 getter를 사용하지 않으려면 출력을 Car 안에 넣어야 하고 이는 Car가 자동차의 정보를 저장하고 레이싱의 결과도 출력하는 두 가지 일을 하게 됩니다. 위에서 설명한 두 가지 조건을 모두 만족하려면 코드를 어떻게 작성해야 하나요?

감사합니다^^

skullkim committed:
study(test/java/study): 학습테스트를 통해 Junit 학습하기 완료

String API, Set collection을 이용해 Junit5 연습 완료

skullkim committed:
study(StringCalculator.java): 문자열 덧셈 계산기 완료

문자열 덧셈 계산기구현을 통해 TDD공부

skullkim committed:
feat(RandomNumber.java): 난수생성기능 완료

자동차 전진 여부를 판단하기 위해 필요한 난수를 생성하는 기능 완료

skullkim committed:
feat(Car.java):자동차 한대의 정보를 담는 클래스 생성

랜덤 메서드를 호출하여 전진 여부를 판단 후 전진하는 메서드 랜덤 메서드를

skullkim committed:
refactor(RandomNumber.java):파일 위치 조정

RandomNumber.java파일을 /utils 내부로 이동

BREKAING CHANGE: RandomNumner.java 파일 위치 수정
as-is:
main/java/racingcar/RandomNumber.java
to-be:
main/java/racingcar/utils/RandomNumber.java

skullkim committed:
feat(InputValidation.java):입력된 데이터 검증 기능 구현

입력한 이름들의 길이가 각 5가 넘지 않도록 검증

skullkim committed:
test(InputValidationTest.java): 테스트 추가

이름 데이터를 검증하는 기능에 대한 테스트 추가

유효하지 않은 이름(6자 이상)이 왔을 때 예외를 던지는지 확인하는 테스트
추가

skullkim committed:
feat(InputValidation.java): 입력된 반복 횟수 검증

입력한 반복 횟수가 정수가 하닌 다른 문자가 포함되 있다면
RunTimeException을 던진다.

skullkim committed:
feat(main/java/views): 입출력 기능 완료

게임 진행을 위해 사용자가 입력해야 하는 자동차 이름과 턴 수를 입력받는
기능 완료. 게임 진행 과정과 결과를 알려주는 기능 완료.
입력이 유효하지 않으면 유효한 입력이 있을 떄 까지 다시 입력받는다.

skullkim committed:
feat(RacingGame.java): 레이싱 게임 로직 구현

이전에 구현한 기능을 활용해 실제 레이싱 게임 로직 구현

skullkim committed:
refactor(src/main/java/racingcar): 불변이어야 하는 변수에 final 추가

불변이여야 하는 지역 변수와 파리미터에 final 추가

skullkim committed:
refactor(Input.java): 에러 문구 출력 기능 추가

잘못된 입력이 있을 때 에러를 알려주는 문구를 출력하는 기능 추가.

BREAK CHANGE: Input class의 inputNames, inputRepeats 메서드 변경
inputNames 메서드:
as-is:
private String inputNames() {
    try {
        ...
    } catch (RuntimeException exception) {
        return inputNames();
    }
}

to-be:
private String inputNames() {
    try {
        ...
    } catch (RuntimeException exception) {
        System.out.println(exception.getMessage() + "\n");
        return inputNames();
    }
}

inputRepeats 메서드:
as-is:
private String inputRepeats() {
    try {
        ...
    } catch (RuntimeException exception) {
        return inputRepeats();
    }
}

to-be:
private String inputRepeats() {
    try {
        ...
    } catch (RuntimeException exception) {
        System.out.println(exception.getMessage() + "\n");
        return inputRepeats();
    }
}

skullkim committed:
refactor(src/main/java.racingcar): 클래스 컨벤션 수정

클래스 시작 부분을 하나의 공백라인으로 시작하게 수정.

skullkim committed:
refactor(src/main/java/racingcar): 게임에 필요한 메시지 통합

게임 진행을 위한 입출력, 에러 안내 메시지를 하나의 클래스에서 관리하기
위해 RacingGameMessage 생성. RacingGameMessage에서 모든 메시지를
관리한다.

skullkim committed:
refactor(Car.java): 상수 추가

전진 여부를 판단하는 숫자와 차 위치 초기값을 상수로 변경.

FORWARD_STANDARD = 4(자동차 전진 여부)
INITIAL_POSITION = 0(초기 차량 위치값)

skullkim committed:
Merge branch 'racing-game' into develop

pkch93 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+	final Input input;
+	final Output output;
+	ArrayList<Car> cars;
+	int repeats;
```
RacingGame의 필드들을 패키지 레벨로 접근제어를 설정하셨는데요. 혹시 이유가 있을까요?

skullkim replied:
실수로 private을 빼먹었네요.. 수정하겠습니다

pkch93 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+
+	final Input input;
+	final Output output;
+	ArrayList<Car> cars;
```
cars가 구체 클래스인 ArrayList로 선언되어있는데요. ArrayList가 List 인터페이스를 구현하고 있어서 List를 참조하도록 만들면 어떨까요? 인터페이스를 참조하도록 만들면 좀 더 유연한 프로그램으로 만들 수 있을거에요.

skullkim replied:
피드백을 토대로 수정했습니다~

pkch93 commented on src/main/java/racingcar/models/Car.java:
```diff
@@ -0,0 +1,33 @@
+
+	final RandomNumber randomNumber;
+	final String name;
+	int position;
```
여기도 RacingGame과 같이 package 레벨로 접근제어자가 정의되어있네요.

skullkim replied:
이것 역시 private으로 수정했습니다

pkch93 commented on src/test/java/racingcar/models/CarTest.java:
```diff
@@ -0,0 +1,39 @@
+@DisplayName("Car 클래스를 테스트한다.")
+class CarTest {
+
+	RandomNumber randomNumber = mock(RandomNumber.class);
```
테스트하기 힘든 부분을 Mocking 하여 테스트를 하셨군요 👍 
Mock 라이브러리를 쓰지 않고 Mocking 할 수 있도록 만들수도 있는데요. 한번 고민해보셔도 좋을거 같아요.

skullkim replied:
mockito를 사용하지 않고 자동차가 전진하는 조건과 전진하지 않는 조건, 이 두 개의 조건에 해당하는 범위의 난수를 각각 생성하는 두 개의 목객체를 만들어 테스트를 진행해 봤습니다. 만든 목 객체는 해당 객체들을 사용한 CarTest와 같은 경로에 있습니다.

pkch93 commented on src/main/java/racingcar/views/Input.java:
```diff
@@ -0,0 +1,54 @@
+			final String names = scanner.next();
+			inputValidation.validateName(names);
+			return names;
+		} catch (RuntimeException exception) {
```
RuntimeException 보다 상황에 적절한 표준예외를 사용하면 어떨까요?
[이펙티브자바 3/e](http://www.kyobobook.co.kr/product/detailViewKor.laf?mallGb=KOR&ejkGb=KOR&barcode=9788966262281) 72장을 참고해보시면 도움이 될거 같아요.

이펙티브자바 3/e 72장을 정리한 블로그 글도 있어서 참고차 첨부드릴께요 https://jaehun2841.github.io/2019/03/10/effective-java-item72/#%EC%84%9C%EB%A1%A0

skullkim replied:
주신 링크를 보고 RuntimeException 보다는 IllegalArgumentException이 더 적절할거 같아서 IllegalArgumentException으로 수정했습니다.

pkch93 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+			.toArray(String[]::new);
+		return String.join(", ", winnerNames);
+	}
+}
```
controller 패키지에 RacingGame이 있는걸로 봐서 컨트롤러로 사용하신걸로 추측되요. 다만, 컨트롤러에 여러 도메인 로직들과 표현 방식(UI)을 가지고 있는데요.. 이 경우 문제점이 어떻게 될까요? 한번 고민해보시고 위 코드를 개선해보면 좋을거 같습니다.

skullkim replied:
컨트롤러와 UI가 하나의 클래스에 존재한다면 단일책임원칙을 위반하게 되는거 같습니다. 이는 캡슐화가 제대로 이뤄지지 않았다는 의미이므로 수정이 필요할거 같습니다. 따라서 해당 메서드를 출력 UI를 담당하는 Output 클래스로 옮겼습니다.

pkch93 commented on src/main/java/racingcar/utils/InputValidation.java:
```diff
@@ -0,0 +1,28 @@
+	}
+
+	public void validateRepeats(final String repeats) {
+		if (!Pattern.matches(NUMBER_PATTERN, repeats)) {
```
Pattern.matches 메서드 문서를 보면 다음과 같은 설명이 있는데요.
```
* If a pattern is to be used multiple times, compiling it once and reusing
* it will be more efficient than invoking this method each time.
```
즉, Pattern을 여러번 사용하는 경우에 컴파일 시켜놓고 재사용하는 것을 권장하고 있습니다. Pattern을 컴파일해서 재사용하는 방향으로 수정해보면 어떨까요?

skullkim replied:
피드백 주신 내용으로 바꾸었습니다

pkch93 commented on src/test/java/racingcar/utils/InputValidationTest.java:
```diff
@@ -0,0 +1,62 @@
+		@ParameterizedTest(name = "{index} {displayName} names={0}")
+		@ValueSource(strings = {"pobi,crong,honux", "pobi"})
+		void checkRightName(final String names) {
+			InputValidation validation = mock(InputValidation.class);
```
따로 stubbing을 하는 부분이 없는데 mock을 사용한 이유가 있을까요?

skullkim replied:
슬랙으로 주신 피드백을 보고 assertDoesNotThrow()를 사용할려 했지만 현재 assertJ를 사용하고 있는 관계로 대신에 assertThatNoeException().isThrownBy()를 사용했습니다.

pkch93 commented on src/test/java/study/SetTest.java:
```diff
@@ -0,0 +1,47 @@
+import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+public class SetTest {
```
학습테스트를 작성하셨군요 👍

pkch93 commented on src/main/java/racingcar/utils/RacingGameMessage.java:
```diff
@@ -0,0 +1,17 @@
+	public static final String CAR_POSITION = "-";
+	public static final String NAME_POSITION_SPLITTER = " : ";
+
+	public final static String CAR_NAME_DISTRIBUTOR = ",";
```
실수하신거 같네요 ㅋㅋ 위 작성해주신 것처럼 키워드들의 순서를 맞춰주시면 좋을거 같습니다.
https://google.github.io/styleguide/javaguide.html#s4.8.7-modifiers

skullkim replied:
실수한 부분 수정했습니다ㅋㅋ.

pkch93 reviewed (changes requested):
안녕하세요 스컬~! 이번에 스컬의 자동차 경주 게임 리뷰어를 맡게된 철시입니다.
우테코에서 첫 PR이신데 고민하신 흔적들이 느껴지네요 ㅎㅎ
몇몇 코멘트들을 남겼는데 확인부탁드릴께요. 혹시 궁금한점 있으시면 코멘트나 DM 주시면 답변 드리겠습니다 ㅎㅎ

---
리뷰 올리시면서 주신 질문들에 대한 제 생각을 남길께요 ㅎㅎ
1. 생성자에서 값을 초기화 할때도 상수를 사용하는 것이 좋나요?
Car객체에 private static final int INITIAL_POSITION = 0;와 같이 자동차 위치를 0이라는 수로 초기화를 할 때 상수로 초기화하지 않아도 코드를 보면 해당 코드가 자동차의 초기 위치를 설정하는 것을 알 수 있는데 이러한 값들도 상수를 지정해주는게 좋은지 알고 싶습니다. (상수 정의의 기준을 알고 싶습니다.)

보통 의미가 불분명한 매직넘버에 의미를 부여하기 위해서 사용을 합니다. 스컬이 생각한대로 이미 그 의미를 알 수 있는 값이기 때문에 굳이 상수로 정의할 필요는 없지 않나 생각합니다.

Car에서 정의한 상수 `FORWARD_STANDARD`와 같은 경우가 상수를 지정하는 좋은 예시가 아닐까 싶어요.

2. getter를 반드시 없애야 하나요?
제가 바로 대답드리는거 보다 [객체지향 생활 체조 규칙 9. 게터/세터/프로퍼티사용 금지](https://developerfarm.wordpress.com/2012/02/01/object_calisthenics_10/)이 무슨 내용인지 한번 확인해보고 고민한 결과를 코멘트나 DM으로 같이 이야기 해보면 좋을거 같습니다 ㅎㅎ

skullkim committed:
refactor(RacingGame.java): 피드백에 따른 상태 수정

상태들의 접근자를 default에서 private으로 변경.
유연석을 위해 ArrayList를 List interface로 변경.

BREAKING CHANGE: 접근자 수정, ArryaList를 List로 변경.
as-is:
public class RacingGame {

    final Input input;
    final Output output;
    ArryaList<Car> cars;
    int repeats;

    ...
}

to-be:
public class RacingGame {

    private final Input input;
    private final Output output;
    private List<Car> cars;
    private int repeats;

    ...
}

skullkim committed:
refactor(Car.java): 상태들의 접근자 수정

피드백에 따라 상태들의 접근자를 default에서 private으로 수정

BREAKING CHANGE: randomNumber, name, position의 접근자 수정
as-is:
public class Car {
    ...

    final RandomNumber randomNumber;
    final String name;
    int position;

    ...
}

to-be:
public class Car {
    ...

    private final RandomNumber randomNumber;
    private final String name;
    private int position;

    ...
}

skullkim committed:
refactor(CarTest.java): RandomNumber 인스턴스 모킹 방식 수정

피드백을 수용해 RandomNumber 인스턴스 모킹 방식을 mockito에서 별도의
목 객체를 만드는 방식으로 수정.

BREAKING CHANGE: RandomNumber의 목 객체 생성, CatTest내의 RandomNumber
인스턴스 모킹 관련 코드 수정.
as-is:
@DisplayName("Car 클래스를 테스트한다.")
class CarTest {

    RandomNumber randomNumber = mock(RandomNumber.class);

    @Test
    @DisplayName("자동차가 앞으로 전진하는지 확인한다")
    void goForward() {
        when(randomNumber.getRandomNumber()).thenReturn(6);
        Car car = new Car(randomNumber, "a");
        ...
    }

    @Test
    @DisplayName("자동차가 앞으로 전진하지 않는지 확인한다")
    void dontGoForward() {
        when(randomNumber.getRandomNumber()).thenReturn(2);
        Car car = new Car(randomNumber, "a");
        ...
    }
}

to-be:
test/java/racingcar/models/RandomNumberMockLargerThanFour.java
public class RandomNumberMockLargerThanFour extends RandomNumber {

    private static final int MAX_EXCLUSIVE = 6;

    @Override
    public int getRandomNumber() {
        Random randomNumber = new Random();
        return randomNumber.nextInt(MAX_EXCLUSIVE) + 4;
    }
}

test/java/racingcar/models/RandomNumberMockLargerThanThree.java
public class RandomNumberMockLargerThanFour extends RandomNumber {

    private static final int MAX_EXCLUSIVE = 6;

    @Override
    public int getRandomNumber() {
        Random randomNumber = new Random();
        return randomNumber.nextInt(MAX_EXCLUSIVE) + 4;
    }
}

@DisplayName("Car 클래스를 테스트한다.")
class CarTest {

    @Test
    @DisplayName("자동차가 앞으로 전진하는지 확인한다")
    void goForward() {
        Car car = new Car(new RandomNumberMockLargerThanFour(), "a");
        ...
    }

    @Test
    @DisplayName("자동차가 앞으로 전진하지 않는지 확인한다")
    void dontGoForward() {
        Car car = new Car(new RandomNumberMockLessThanThree(), "a");
        ...
    }
}

skullkim committed:
refactor(RacingGameMessage.java): 키워드 순서 변경

코딩 컨벤션에 맞춰 CAR_NAME_DISTRIBUTOR의 키워드 순서변경

BREAKING CHANGE: CAR_NAME_DISTRIBUTOR의 static과 final 키워드 순서 변경
as-is:
public final static String CAR_NAME_DISTRIBUTOR = ",";

to-be:
public static final String CAR_NAME_DISTRIBUTOR = ",";

skullkim committed:
refactor(InputValidation.java): Pattern객체 사용 방식 변경

피드백을 수용해 성능 향상을 위해 Pattern을 컴파일 한 후 사용하게 수정

BREAKING CHNAGE: Pattern으로 입력값과 정규식 일치 여부를 판단하는 코드
수정
as-is:
public class InputValidation {

    ...
    private final static String NUMBER_PATTERN = "^[0-9]*$";

    ...

    public void validateRepeats(final String repeats) {
        if (!COMPILED_NUMBER_PATTERN.matcher(repeats).matches()) {
            ...
        }
    }
}

to-be:
public class InputValidation {

    ...
    private final static String NUMBER_PATTERN = "^[0-9]*$";
    private final static Pattern COMPILED_NUMBER_PATTERN = Pattern.compile(NUMBER_PATTERN);

    ...

    public void validateRepeats(final String repeats) {
        if (!COMPILED_NUMBER_PATTERN.matcher(repeats).matches()) {
            ...
        }
    }
}

skullkim committed:
refactor(RacingGame.java):우승자 이름을 합치는 부분을 다른 클래스로 이동

우승자 이름을 합치는 것은 view를 담당하는 객체가 해야할 역할이다. 따라서
이 역할은 컨트롤러인 RacingGame에서 output view를 담당하는 Output으로
이동.

BREAKING CHANGE: RacingGame에서 우승자 이름을 합치는 부분은 Output으로
이동.
RacingGame:
as-is:
public class RacingGame {

    ...
    private final Output output;
    private List<Car> cars;

    ...

    public void startGame() {
        inputValidData();
        startRacing();
    }

   ...

    private void startRacing() {
        ...
        output.printWinner(findWinner());
    }

    private String findWinner() {
        final int farthestPosition = findFarthestPosition();
        final List<Car> winners = getWinners(farthestPosition);
        return joinWinnerNames(winners);
    }

   ...

    private List<Car> getWinners(final int farthestPosition) {
        return cars.stream()
            .filter((car) -> farthestPosition == car.getPosition())
            .collect(Collectors.toList());
    }

    private String joinWinnerNames(final List<Car> winners) {
        final String[] winnerNames = winners.stream()
            .map(Car::getName)
            .toArray(String[]::new);
        return String.join(", ", winnerNames);
    }
}

to-be:
public class RacingGame {

    ...
    private final Output output;
    private List<Car> cars;

    ...

    public void startGame() {
        inputValidData();
        startRacing();
    }

   ...

    private void startRacing() {
        ...
        output.printWinner(findWinner());
    }

    private List<Car> findWinner() {
        final int farthestPosition = findFarthestPosition();
        return getWinners(farthestPosition);
    }

   ...

    private List<Car> getWinners(final int farthestPosition) {
        return cars.stream()
            .filter((car) -> farthestPosition == car.getPosition())
            .collect(Collectors.toList());
    }
}

Output:
as-is:
public class Output {

    ...

    public void printWinner(final String winners) {
        System.out.print(winners + RacingGameMessage.WINNER);
    }
}

to-be:
public class Output {

    ...

    public void printWinner(final List<Car> winners) {
        System.out.print(joinWinnerNames(winners) + RacingGameMessage.WINNER);
    }

    private String joinWinnerNames(final List<Car> winners) {
        final String[] winnerNames = winners.stream()
            .map(Car::getName)
            .toArray(String[]::new);
        return String.join(", ", winnerNames);
    }
}

skullkim committed:
refactor(InputValidation.java, Input.java): 예외 유형 변경

피드백을 수정해 입력 값이 유효하지 않을 경우 사용하는 예외 변경

BREAKING CHANGE: RunTimeException을 IllegalArgumentException으로 변경
InputValidation:
as-is:
public class InputValidation {
    ...

    private void checkNameLength(final String carName) {
        if (carName.length() > MAX_NAME_LENGTH) {
            throw new RuntimeException(RacingGameMessage.NAME_LENGTH_ERROR);
        }
    }

    public void validateRepeats(final String repeats) {
        if (!COMPILED_NUMBER_PATTERN.matcher(repeats).matches()) {
            throw new RuntimeException(RacingGameMessage.REPEATS_ERROR);
        }
    }
}

to-be:
public class InputValidation {

    ...

    private void checkNameLength(final String carName) {
        if (carName.length() > MAX_NAME_LENGTH) {
            throw new IllegalArgumentException(RacingGameMessage.NAME_LENGTH_ERROR);
        }
    }

    public void validateRepeats(final String repeats) {
        if (!COMPILED_NUMBER_PATTERN.matcher(repeats).matches()) {
            throw new IllegalArgumentException(RacingGameMessage.REPEATS_ERROR);
        }
    }
}

as-is:
public class Input {

    ...

    private String inputNames() {
        try {
            ...
        } catch (RuntimeException exception) {
            ...
        }
    }

    ...

    private String inputRepeats() {
        try {
            ...
        } catch (RuntimeException exception) {
            ...
        }
    }
}

to-be:
public class Input {

    ...

    private String inputNames() {
        try {
            ...
        } catch (IllegalArgumentException exception) {
            ...
        }
    }

    ...

    private String inputRepeats() {
        try {
            ...
        } catch (IllegalArgumentException exception) {
            ...
        }
    }
}

skullkim committed:
chore(InputValidationTest.java): 누락된 파일 추가

이전 커밋에 누락된 테스트 파일 추가

skullkim committed:
refactor(InputValidationTest.java): 정상 실행 테스트 코드 수정

올바른 입력 데이터를 검증할 떄 에러를 발생하시키지 않는지 테스트 하는
유닛 테스트 수정

BREAKING CHANGE: mock을 사용해 verify(호출 여부)를 제크하는 로직에서
에러가 발생하지 않음을 체크하는 테스트로 변경

checkRightName 메서드:
as-is:
void checkRightName(final String names) {
    InputValidation validation = mock(InputValidation.class);
    validation.validateName(names);
    verify(validation).validateName(names);
}

to-be:
void checkRightName(final String names) {
    InputValidation validation = new InputValidation();
    assertThatNoException().isThrownBy(() -> validation.validateName(names));
}

checkRightRepeats 메서드:
as-is:
void checkRightRepeats(final String repeats) {
    InputValidation validation = mock(InputValidation.class);
    validation.validateRepeats(repeats);
    verify(validation).validateRepeats(repeats);
}

to-be:
void checkRightRepeats(final String repeats) {
    InputValidation validation = new InputValidation();
    assertThatNoException().isThrownBy(() -> validation.validateName(repeats));
}

skullkim committed:
refactor(InputValidation.java): 중복 코드 제거

테스트에서 중복된 코드 제거

BREAKING CHANGE: 유닛 테스트 마다 존재했던 InputValidation 인스턴스 제거

as-is:
@DisplayName("InputValidation클래스 테스트")
class InputValidationTest {

    ...
    class NameTest {

        ...
        void checkRightName(final String names) {
            InputValidation validation = new InputValidation();
            assertThatNoException().isThrownBy(() -> validation.validateName(names));
        }

        ...
        void checkWrongName(final String names) {
            InputValidation validation = new InputValidation();
            assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> validation.validateName(names))
                .withMessageMatching("이름은 5자리 이하만 가능합니다.");
        }
    }

    ...
    class RepeatTest {

        ...
        void checkRightRepeats(final String repeats) {
            InputValidation validation = new InputValidation();
            assertThatNoException().isThrownBy(() -> validation.validateName(repeats));
        }

        ...
        void checkWrongRepeats(final String repeats) {
            InputValidation validation = new InputValidation();
            assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> validation.validateRepeats(repeats))
                .withMessageMatching("반복횟수는 정수만 입력 가능합니다.");
        }
    }
}

to-be:
@DisplayName("InputValidation클래스 테스트")
class InputValidationTest {

    InputValidation validation = new InputValidation();

    ...
    class NameTest {

        ...
        void checkRightName(final String names) {
            assertThatNoException().isThrownBy(() -> validation.validateName(names));
        }

        ...
        void checkWrongName(final String names) {
            assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> validation.validateName(names))
                .withMessageMatching("이름은 5자리 이하만 가능합니다.");
        }
    }

    ...
    class RepeatTest {

        ...
        void checkRightRepeats(final String repeats) {
            assertThatNoException().isThrownBy(() -> validation.validateName(repeats));
        }

        ...
        void checkWrongRepeats(final String repeats) {
            assertThatExceptionOfType(IllegalArgumentException.class)
                .isThrownBy(() -> validation.validateRepeats(repeats))
                .withMessageMatching("반복횟수는 정수만 입력 가능합니다.");
        }
    }
}

pkch93 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,69 @@
+import racingcar.views.Input;
+import racingcar.views.Output;
+
+public class RacingGame {
```
UI에 대한 부분은 View로 책임을 전달 주셨네요 👍 
다만 몇몇 도메인 로직 아직 컨트롤러 역할을 하는 RacingGame에 있는걸로 보이는데요. Controller에 도메인 로직이 있을때 어떤 장단점이 있을지를 생각해보시면서 다음단계를 진행하시면 좋을거 같습니다 :)

pkch93 reviewed (approved):
주말임에도 피드백 반영하느라 고생많으셨어요..!
다음 단계에서 뵐께요 :)
